### ISSUE TITLE
Does not pick up messages defined in a `format*` method

### ISSUE THREAD

**Issue by baer** [formatjs/babel-plugin-react-intl#18]
This should probably be behind a config since it could cause a lot of bad data if not used carefully but so far this plugin only picks up messages in a `defineMessages` block and in the react components. It would be nice to be able to use `format*()` for a11y without having to revert to `defineMessages`.

Happy to PR this :smile:

**Comment by ericf** [comment]
The current design of the plugin works off of React Intl `export`s, which is why `defineMessages()` was created. By limiting the extractions to things React Intl exports, it greatly simplifies the plugin and reduces the reliance on heuristics and pattern matching.

In some apps, I'm seeing people prefer to use `defineMessages()` at the top (or bottom) of their modules, even if they are only using `<FormattedMessage>`. This is similar to what you see with styles in React Native.

**Comment by baer** [comment]
I suppose it could get hairy trying to figure out which `format*()` calls belonged to react-intl but it doesn't seem insurmountable. The problem is mainly that you have an inconsistent API. One message format will auto extract and the other will not. 

What if we flip this around and have an option to disallow (or at least not extract) messages defined in `<FormattedMessage>` components? That way, for large projects you can enforce a single way to do things.

**Comment by ericf** [comment]
> I suppose it could get hairy trying to figure out which format*() calls belonged to react-intl but it doesn't seem insurmountable.

One problem it's possible to use `formatMessage()` without importing `'react-intl'`. I'm suggesting that people use `injectIntl()` which would be a good signal, but someone could wrap a component from the outside. This would remove the filtering function that only looks at files that import `react-intl`.

> The problem is mainly that you have an inconsistent API.

Yeah. I tried to make it work by having all the `format*()` functions be exported from React Intl, but it was more error prone and seemed less economic.

> What if we flip this around and have an option to disallow (or at least not extract) messages defined in <FormattedMessage> components? That way, for large projects you can enforce a single way to do things.

I think this is worth exploring more too; and is related to #17.

**Comment by baer** [comment]
Sounds like you've been down this road. I've been looking the packages to get a feel for how different parts of the tooling work and I have to compliment your code. It's just... so clean! :clap: thanks again for putting so much into this!

As for disallowing messages defined in the react components - I'll see if I can get a PR to you either tomorrow of Monday.

**Comment by ericf** [comment]
Thanks! Sounds good.

**Comment by iammerrick** [comment]
@ericf What is the alternative for imperatively using messages if we can't use injectIntl + formatMessage. We should injectIntl and call defineMessages and that will get picked up via this plugin? I guess what I'm asking is how do we refactor this: 

```
        <InputWithIcon
          onInputChange={this.handleSearchChange}
          icon={<Icon icon={Icon.OPTIONS.SEARCH} />}
          placeholder={this.props.intl.formatMessage({
            id: 'AllPeopleHandler__SearchPeople',
            defaultMessage: 'Search people',
            description: 'Input placeholder for searching all people.'
          })}
        />
```

To support being extracted via this babel plugin?

**Comment by iammerrick** [comment]
When I look at defineMessages: https://github.com/yahoo/react-intl/blob/master/src/define-messages.js It seems like an identity function, so do we just wrap the object literals passed to formatMessage(...) with defineMessages? `this.props.intl.formatMessage(defineMessages(...))` ?

**Comment by iammerrick** [comment]
what would define messages look like, is the key the locale? Isn't the default locale implicit via defaultMessages?

**Comment by ericf** [comment]
@iammerrick I think of this similar to the systems that allow you to define your CSS styles in your React components:

```js
import React, {Component} from 'react';
import {defineMessages, injectIntl} from 'react-intl';
import Icon from './icon';
import InputWithIcon from './input-with-icon';

const messages = defineMessages({
  searchPeople: {
    id: 'AllPeopleHandler__SearchPeople',
    defaultMessage: 'Search people',
    description: 'Input placeholder for searching all people.'
  }
});

class MyComponent extends Component {
  componentDidMount() {
    this.handleSearchChange = this.handleSearchChange.bind(this);
  }

  handleSearchChange() {}

  render() {
    return (
      <InputWithIcon
        onInputChange={this.handleSearchChange}
        icon={<Icon icon={Icon.OPTIONS.SEARCH} />}
        placeholder={this.props.intl.formatMessage(messages.searchPeople)}
      />
    );
  }
}

export default injectIntl(MyComponent);
```

The main idea here is that the plugin works on `import`s, and `defineMessages` is an import and static, while `this.props.intl.formatMessage()` is dynamic and would have to rely on heuristics to properly find its use cases to extract messages. I've actually found that some teams like to _always_ use `defineMessages()` in their components.

**Comment by iammerrick** [comment]
This is great information! Thank you! May I recommend an example of this that doesn't use local as the key? :-) This is awesome. By the way, this is the most wonderful comprehensive i18n solution I've ever worked with!

**Comment by plandem** [comment]
so...if using formatMessage  then it will not be extracted? Am i correct?

because each my component looks like:
```
import {defineMessages, injectIntl } from 'react-intl';

const messages = defineMessages({
  dashboardSaved: {
    id: 'app.dashboard-builder.dashboard-saved',
    defaultMessage: 'Dashboard was successfully saved'
  },
  buttonLogout: {
    id: 'app.button-logout',
    defaultMessage: 'Logout'
  }
});

  onSaveHandler(e) {
    const { actions, intl: { formatMessage } } = this.props;

    this.setState({ isDisabled: true });
    
    actions.main.save().then(() => {
      this.setState({ isDisabled: false });
      message.success(formatMessage(messages.dashboardSaved));
    }).catch((error) => {
      this.setState({ isDisabled: false });
      message.error(error);
    });
  }
```

**Comment by ericf** [comment]
They will be picked up because you're using `defineMessages`.

**Comment by plandem** [comment]
ok, but after:
```
babel --presets es2015,react,stage-0 src >1 
```
i have just one big file (looks like concated 'src') and that's all. I don't see any extracted messages

**Comment by ericf** [comment]
@plandem it's in the README: https://github.com/yahoo/babel-plugin-react-intl#usage

**Comment by jardakotesovec** [comment]
My main struggle with using defineMessages at the beginning of file or in separate file (per component) is that it gets easy out of sync with actual component over time - meaning that I will have some translations that are not used anymore in the component.   

Thats reason why I think using `format*` methods inline is handy.  I am wondering how do you deal with this issue as consensus here seems to be keep messages separately from component.

**Comment by ericf** [comment]
It would be a bit of boilerplate but you could use object destructuring assignments which a tool like ESLint could warn about unused variables.

```js
const {foo, bar} = defineMessages({
  foo: {
    id: 'FOO',
    defaultMessage: 'foo',
  },
  bar: {
    id: 'BAR',
    defaultMessage: 'bar',
  },
});
```

**Comment by plandem** [comment]
also it will warn about duplicates

**Comment by JasonKleban** [comment]
I am also confused about how all the pieces are supposed to fit together.  Here's what I have so far:

src/messages.js:
```
import {defineMessages} from 'react-intl';

const messages = defineMessages({
    hello: {
        id: 'app.hello',
        description: 'Message to greet the user.',
        defaultMessage: 'Hello, {name}!'
    },
    goodbye: {
        id: 'app.goodbye',
        description: 'Message to dismiss the user.',
        defaultMessage: 'Goodbye, {name}!'
    }
}
```

src/components/my-component.js
```
    ...
    <FormattedMessage 
        id="app.my-message" 
        description="n Apple(s)"
        defaultMessage="{numOfApples, number} {numOfApples, plural, one {Apple} other {Apples}}" 
        values={{ numOfApples : data.numOfApples }} />
    ...
```

babel-plugin-react-intl seems to be configured correctly in .babelrc because I get:

dist/messages.json:
```
[
  {
    "id": "app.hello",
    "description": "Message to greet the user.",
    "defaultMessage": "Hello, {name}!"
  },
  {
    "id": "app.goodbye",
    "description": "Message to dismiss the user.",
    "defaultMessage": "Goodbye, {name}!"
  }
]
```

dist/components/my-component.json:
```
[
  {
    "id": "app.my-message",
    "description": "n Apple(s)",
    "defaultMessage": "{numOfApples, number} {numOfApples, plural, one {Apple} other {Apples}}" 
  },
  ...
]
```

And I have react-intl setup correctly I think:

```
<IntlProvider locale={theLocale} messages={???}>
    ...
</IntlProvider>
```

But I don't know what to put in the messages prop, because it wants an object with members indexed on the message id, but the json files are arrays of objects containing an ID instead.  Passing the json file data to defineMessages() seems to have no effect, so I don't think that's right.

```
type IntlConfig = {
    locale: string,
    formats: object,
    messages: {[id: string]: string},

    defaultLocale: string = 'en',
    defaultFormats: object = {},
};
```

1. How do we convert the json data file of messages to the format expected by IntlConfig?
2. Are we to manually create `messages.[culture].json` files (or some naming/organizational convention?) or is there some built-in feature?  Is there a prescribed convention for naming and organizing the per-culture translations?
3. What is the best practice for downloading to the client only the translation files needed for a particular culture?  Is there a built-in feature that does this automatically?
4. Once we have the necessary per-culture translation json files loaded, what is the recommended way to register them?  Do we combine them somehow?
5. Is there a way to refer to an already defined message by id in a `<FormatMessage />`?  I understand that we could refer to a single declaration of a `<FormatMessage />` by some predefined variable and reuse that, but I wouldn't want to reinvent a way of dealing with that across sourcecode .js files if a built-in feature for this already exists.

Thanks!

**Comment by web2style** [comment]
Please, consider adding information about this in the README. This can save somebody a day.

**Comment by peterp** [comment]
@ericf I've read the readme a few times over... and I'm not exactly sure why I don't get a defined bunch of messages? Do I invoke babel-cli manually? Should reloading my app/ webpage invoke babel and therefore generate the messages?

**Comment by longlho** [comment]
fixed by #165
